Program.cs

```diff
@@ -123,173 +102,52 @@ public static void Main(string[] args)
-                    Directory.CreateDirectory(stagingTarget);
-                    foreach (var file in Directory.GetFiles(Path.Combine(templatesPath, app_dir)))
-                        new FileInfo(file).CopyTo(Path.Combine(stagingTarget, Path.GetFileName(file)));
-                    runCommand("chmod", $"+x {stagingTarget}/AppRun");
-                    publishDir = Path.Combine(stagingTarget, "usr/bin/");
-                    extraCmd += $" -p \"{stagingTarget}\"";
-                }
+                case RuntimeInfo.Platform.Windows:
+                    builder = new WindowsBuilder(version, getArg(0));
+                    break;
 
-                runCommand("dotnet", $"publish -f net8.0 -r {rid} {ProjectName} -o \"{publishDir}\" --configuration Release /p:Version={version} --self-contained");
+                case RuntimeInfo.Platform.Linux:
+                    builder = new LinuxBuilder(version);
+                    break;
 
-                if (targetPlatform == RuntimeInfo.Platform.Windows)
-                {
-                    bool rcEditCommand = runCommand("tools/rcedit-x64.exe", $"\"{Path.Combine(publishDir, "osu!.exe")}\" --set-icon \"{iconPath}\"", exitOnFail: false);
-                    if (!rcEditCommand)
-                    {
-                        // Retry again with wine
-                        // TODO: Should probably change this to use RuntimeInfo.OS checks instead of fail values
-                        bool wineRcEditCommand = runCommand("wine", $"\"{Path.GetFullPath("tools/rcedit-x64.exe")}\" \"{Path.Combine(publishDir, "osu!.exe")}\" --set-icon \"{iconPath}\"",
-                            exitOnFail: false);
-                        if (!wineRcEditCommand)
-                            error("Failed to set icon on osu!.exe");
-                    }
-
-                    if (!string.IsNullOrEmpty(WindowsCodeSigningCertPath))
-                    {
-                        string? codeSigningPassword;
-
-                        if (args.Length > 0)
-                        {
-                            codeSigningPassword = args[0];
-                        }
-                        else
-                        {
-                            Console.Write("Enter code signing password: ");
-                            codeSigningPassword = readLineMasked();
-                        }
-
-                        extraCmd += string.IsNullOrEmpty(codeSigningPassword)
-                            ? ""
-                            : $" --signParams=\"/td sha256 /fd sha256 /f {WindowsCodeSigningCertPath} /p {codeSigningPassword} /tr http://timestamp.comodoca.com\"";
-                    }
-
-                    extraCmd += $" --splashImage=\"{splashImagePath}\" --icon=\"{iconPath}\" -p {stagingPath}";
-                }
+                case RuntimeInfo.Platform.macOS:
+                    builder = new MacOSBuilder(version, getArg(3));
+                    break;

```

Comment:
This should probably do checks for the host OS before creating the `MacOSBuilder`

Velopack supports cross compilation, but only in certain configurations. MacOS binaries can only be built on MacOS, it's best to check for that here and add `SupportedOSPlatformAttribute` to `MacOSBuilder`

See https://docs.velopack.io/packaging/cross-compiling#supported-configurations

Reply:
I don't think there's a reason to do this. It will error out with a readable exception, which is the best we can ever do.

Original code also had no safeguards, even though it was doing the same stuff.